kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/SchemaUtils.java

```diff
@@ -648,4 +651,47 @@ protected static void validateSupportedType(DataType dataType) {
       throw unsupportedDataType(dataType);
     }
   }
+
+  /**
+   * If IcebergWriterCompatV1 is enabled, we need to ensure that map struct keys don't change. This
+   * validates that
+   */
+  private static void validateNoMapStructKeyChanges(
+      StructType currentSchema, StructType newSchema, Map<String, String> configuration) {
+    if (TableConfig.ICEBERG_WRITER_COMPAT_V1_ENABLED.fromMetadata(configuration)) {
+      // do the check if the feature is enabled now
+      List<Tuple2<List<String>, StructField>> columnPathToMaps =
+          filterRecursively(
+              currentSchema,
+              true, /* recurseIntoMapOrArrayElements, need to check maps inside arrays */
+              false, /* stopOnFirstMatch */
+              sf -> {
+                DataType dataType = sf.getDataType();
+                return (dataType instanceof MapType)
+                    && ((MapType) dataType).getKeyField().getDataType() instanceof StructType;
+              });
+      Map<Integer, StructField> updatedFieldsById = fieldsById(newSchema);
+      for (Tuple2<List<String>, StructField> pathAndField : columnPathToMaps) {
```

Comment:
nit - We could probably nest all of this within the `filterRecursively` fx? Then the type checks and casts/code that uses them can be together

Reply:
I don't use `filterResusively` anymore as I'm using `SchemaChanges`